chatgpt-shell-openai.el

```diff
@@ -268,34 +263,35 @@ or
 
 (cl-defun chatgpt-shell-openai--make-payload (&key model context settings)
   "Create the API payload using MODEL CONTEXT and SETTINGS."
-  (chatgpt-shell-openai-make-chatgpt-request-data
+  (apply
+   #'chatgpt-shell-openai-make-chatgpt-request-data
    :system-prompt (map-elt settings :system-prompt)
    :context context
    :version (map-elt model :version)
    :temperature (map-elt settings :temperature)
-   :streaming (map-elt settings :streaming)))
+   :streaming (map-elt settings :streaming)
+   :other-params (map-elt model :other-params)))
 
-(cl-defun chatgpt-shell-openai--handle-chatgpt-command (&key model command context shell settings)
+(cl-defun chatgpt-shell-openai--handle-chatgpt-command (&key model command context shell settings (key #'chatgpt-shell-openai-key) (filter #'chatgpt-shell-openai--filter-output))
   "Handle ChatGPT COMMAND (prompt) using MODEL, CONTEXT, SHELL, and SETTINGS."
-  (unless (chatgpt-shell-openai-key)
+  (unless (funcall key)
     (funcall (map-elt shell :write-output) "Your chatgpt-shell-openai-key is missing")
     (funcall (map-elt shell :finish-output) nil))
   (shell-maker-make-http-request
    :async t
-   :url (concat chatgpt-shell-api-url-base
-                (or (map-elt model :path)
-                    (error "Model :path not found")))
+   :url (chatgpt-shell-openai--make-url :model model)
    :proxy chatgpt-shell-proxy
    :data (chatgpt-shell-openai-make-chatgpt-request-data
           :prompt command
           :system-prompt (map-elt settings :system-prompt)
           :context context
           :version (map-elt model :version)
           :temperature (map-elt settings :temperature)
-          :streaming (map-elt settings :streaming))
+          :streaming (map-elt settings :streaming)
+          :other-params (map-elt model :other-params))
    :headers (list "Content-Type: application/json; charset=utf-8"
-                  (format "Authorization: Bearer %s" (chatgpt-shell-openai-key)))
-   :filter #'chatgpt-shell-openai--filter-output
+                  (format "Authorization: Bearer %s" (funcall key)))
```

Comment:
(chatgpt-shell-openai-key) used to support either strings or functions in var chatgpt-shell-openai-key. If I'm understanding correctly, we lose this behavior and will break for users setting chatgpt-shell-openai-key to a string?

Reply:
Yeah, this should be easy to fix. We can just bind chatgpt-shell-openai-key to key via let.

Reply:
Actually, I think this isn't a problem. There's a `chatgpt-shell-openrouter-key` function which is analogous to `chatgpt-shell-openai-key`. When the key parameter is passed it is one of these and they both handle the case when the variable of the same name is a function.

Reply:
You're right. Working. Tried it out.